File: backend/app/schemas/mcp.py
"""
MCP 服务器 API Schema 定义。
"""
from typing import Optional, Any
from datetime import datetime
from pydantic import BaseModel, Field


class MCPServerCreate(BaseModel):
    """创建 MCP Server 请求。"""
    name: str = Field(..., min_length=1, max_length=100, description="服务器名称")
    type: str = Field(default="stdio", pattern="^(http|stdio)$", description="传输类型：http 或 stdio")
    config: dict = Field(default_factory=dict, description="配置信息，如 url、command、args、env 等")
    status: str = Field(default="connected", description="连接状态")


class MCPServerUpdate(BaseModel):
    """更新 MCP Server 请求。"""
    name: Optional[str] = Field(None, max_length=100)
    type: Optional[str] = Field(None, pattern="^(http|stdio)$")
    config: Optional[dict] = None
    status: Optional[str] = None


class MCPServerOut(BaseModel):
    """MCP Server 输出。"""
    id: str
    name: str
    type: str
    config: Optional[Any] = None  # JSON
    status: str
    user_id: str
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class MCPServerTestRequest(BaseModel):
    """测试 MCP Server 连接请求。"""
    type: str = Field(default="stdio", pattern="^(http|stdio)$")
    config: dict = Field(default_factory=dict)


class MCPServerTestResult(BaseModel):
    """测试 MCP Server 连接结果。"""
    success: bool
    message: str
    tool_count: int = 0
    tools: list[dict] = Field(default_factory=list)

File: frontend/src/api/mcp.ts
import apiClient from './client'

// HTTP 模式 + Stdio 模式统一接口
export interface MCPServerItem {
  id: string
  name: string
  type: string            // http | stdio
  config: Record<string, any> | null
  status: string          // connected, disconnected, error, connecting
  user_id: string
  created_at: string | null
  updated_at: string | null
}

export interface MCPStdioServerItem {
  name: string
  command: string
  args: string[]
  status: string            // running, stopped, crashed, failed
  tool_count: number
  started_at: string | null
  restart_count: number
}

export interface MCPToolItem {
  name: string
  display_name: string
  description: string
}

export interface MCPToolCallRequest {
  tool_name: string
  arguments: Record<string, any>
}

// 创建/更新 Server 的请求
export interface MCPServerCreateRequest {
  name: string
  type: string        // "http" | "stdio"
  config: Record<string, any>
  status?: string
}

export interface MCPServerTestRequest {
  type: string
  config: Record<string, any>
}

export interface MCPServerTestResult {
  success: boolean
  message: string
  tool_count: number
  tools: Record<string, any>[]
}

export const mcpApi = {
  // ── DB-backed CRUD（统一管理 HTTP + Stdio） ──
  listServers() {
    return apiClient.get<MCPServerItem[]>('/mcp/servers')
  },

  createServer(data: MCPServerCreateRequest) {
    return apiClient.post<MCPServerItem>('/mcp/servers', data)
  },

  getServer(serverId: string) {
    return apiClient.get<MCPServerItem>(`/mcp/servers/${serverId}`)
  },

  updateServer(serverId: string, data: Partial<MCPServerCreateRequest>) {
    return apiClient.put<MCPServerItem>(`/mcp/servers/${serverId}`, data)
  },

  deleteServer(serverId: string) {
    return apiClient.delete(`/mcp/servers/${serverId}`)
  },

  // ── 连接测试 ──
  testConnection(data: MCPServerTestRequest) {
    return apiClient.post<MCPServerTestResult>('/mcp/servers/test', data)
  },

  // ── Stdio 进程管理 ──
  connectStdio(serverId: string) {
    return apiClient.post<MCPServerItem>(`/mcp/servers/stdio?server_id=${serverId}`)
  },

  disconnectStdio(serverName: string) {
    return apiClient.delete(`/mcp/servers/stdio/${serverName}`)
  },

  // ── 旧版兼容 ──
  connect(server_name: string, server_url: string) {
    return apiClient.post<MCPToolItem[]>('/mcp/servers', {
      name: server_name,
      type: 'http',
      config: { url: server_url },
    })
  },

  disconnect(server_name: string) {
    return apiClient.delete(`/mcp/servers/${server_name}`)
  },

  listStdioServers() {
    return apiClient.get<MCPStdioServerItem[]>('/mcp/servers/stdio')
  },

  getStdioServer(server_name: string) {
    return apiClient.get<MCPStdioServerItem>(`/mcp/servers/stdio/${server_name}`)
  },

  loadConfig() {
    return apiClient.post('/mcp/load-config')
  },

  // ── 缓存管理 ──
  clearCache(sessionId?: string) {
    const params = sessionId ? `?session_id=${sessionId}` : ''
    return apiClient.post(`/mcp/cache/clear${params}`)
  },
}